Implement yaml

JayOrten opened the pull request:
Note 1/27: I've tested all of these changes...except train and gridsearch fully, because I couldn't get a single gpu to test on. We can hold off until we've had a chance to fully test those scripts, but any needed changes would be minor.

This PR implements YAML configuration files, intended to be used project wide for any configuration needs.

The design choice of the YAML file is intended to eliminate repetition of commonly used parameters across code, as well as simplify future changes and refactors. The power of the YAML file is that it is very flexible: a developer can add to it as they see fit for whatever parameters they need, with the knowledge that the user will be aware of what parameters are needed when they run the code.

A few important notes about this design:
- The YAML is intended to store any path information. Absolute paths work best. The idea is that the YAML allows us to decouple any path information or logic from the code; therefore, when the user is executing the code, they can just look into the YAML to understand all of the paths needed/being used. This brings the flexibility that a custom YAML can work with any directory configuration, which is especially useful on the compute clusters. The downside is that the YAML may acquire multiple, somewhat repetitious paths. However, each path is clearly seen, and should serve a distinct purpose across multiple scripts.

- The YAML file eliminates repetition across code. The YAML is intended to deliver parameters straight to where they are used. That way, when code needs to be changed, only the YAML and the place where the parameter is used needs to be changed. The downside of this is that each script is only taking the config object as an argument, and using that everywhere. While being less transparent, this pattern is fairly common and allows more flexibility both during development and use.

- If there is a mistake in any parameter configuration, it is very easy to track it down in the code, straight to the YAML. If a parameter in the YAML is missing, it will be easily seen. This design also assumes no default arguments, which makes for a cleaner workflow, as any parameters being used are stored in the single YAML file.

The major contributions/changes of this PR are:

- Removes any argument parsing from major scripts, and replaces them with YAML -> config object logic. Adds a struct class to work with this
- Adds a configuration folder with default config, as well as a user folder to contain custom configs, to be ignored by git.
- Refactors/renames many common parameters and paths across scripts to be uniform
- Changes all scripts in scripts folder to be standard and using YAML. Adds an additional user_scripts folder for custom scripts, to be ignored by git.

RLin8103 committed:
Updated YAML configuration for train_model.py

jtappen1 committed:
Created Yaml class struct

jtappen1 committed:
Merge remote-tracking branch 'origin/hugging_faceify' into implement_yaml

Merging HF and YAML Branches

jtappen1 committed:
Added yaml functionality for train_tokenizer.py, tokenize_data.py, and download_data.py

jtappen1 committed:
Added some inital files for train_model.py to include yaml

jtappen1 committed:
Merge remote-tracking branch 'origin/main' into implement_yaml

jtappen1 committed:
Merged main to implement Yaml and tested download_data, train_tokenizer, and tokenize data and they all work.

jtappen1 committed:
Completed YAML cleanup

JayOrten committed:
Minor changes to get train running

JayOrten committed:
Reorganize yaml file and add user folder

JayOrten committed:
Organize and cleanup imports

JayOrten committed:
refactor grid search to use yaml

JayOrten committed:
Add comments to yaml

JayOrten committed:
Directly pass config in train

JayOrten committed:
Refactor other scripts to directly use config

JayOrten committed:
update and standardize scripts, minor fixes

JayOrten committed:
minor fixes to get train working

JayOrten committed:
remove random weights folder

KimballNJardine commented on configs/template_config.yaml:
```diff
@@ -0,0 +1,94 @@
+seq_len: 128
+# value_embed_dim (int): Value embed dimension size.
+value_embed_dim: 12
+# vocab_size (int): Maximum vocabulary size (number of unique tokens in vocabulary.
```
Missing ")" at end of line.

KimballNJardine commented on configs/template_config.yaml:
```diff
@@ -0,0 +1,94 @@
+# Tensorboard path
+tboard_path: ~
+
+splits:
```
An explanation of splits would be good: Example:
\# Splits
\# Fraction Train, Fraction Validation, Fraction Test

KimballNJardine commented on configs/template_config.yaml:
```diff
@@ -0,0 +1,94 @@
+#        The YAML config file is then passed as an argument to any scripts.
+#
+#        For example, to train a model, run:
+#        python train_model.py ./configs/my_config.yaml
```
I am not certain, so correct me if I am wrong, but would it be a good idea to suggest the following instead?:

python train_model.py ./configs/user_configs/my_config.yaml.

If I understand the new gitignore file, that folder is where everyone should put their YAML files.

JayOrten replied:
Good catch

KimballNJardine commented on download_data.py:
```diff
@@ -19,22 +18,25 @@ def download_data(
-        filename = dataset_subset + ".parquet"
+        filename = config.dataset_subset + ".parquet"
+        dataset.to_parquet(dataset_dir / filename)
+        filename = config.dataset_subset + ".parquet"
```
These lines are duplicated. Please remove duplicate lines 37 and 38.

KimballNJardine commented on download_data.py:
```diff
@@ -44,16 +46,12 @@ def download_data(
-        help="Path to directory in which Hugging Face datasets are downloaded.")
+    with open(config_path, 'r') as f:
+        config = yaml.safe_load(f)
 
```
The white space here is excessive, in my opinion. This is repeated in the other files though, so it is consistent. Let's decide on if to keep all the blank lines or condense them before approving the pull request.

KimballNJardine commented on grid_search.py:
```diff
@@ -148,21 +129,12 @@ def grid_search(
-        help="Maximum number of unique tokens in vocabulary.")
+    args = sys.argv
+    config_path =args[1]
+
```
See earlier comment about white space.

JayOrten replied:
discussed

KimballNJardine commented on scripts/download_data.sh:
We may want to rename the scripts folder "generic_scripts", or something like that. That is not a requirement, but something to discuss.

JayOrten replied:
discussed

KimballNJardine commented on tokenize_data.py:
```diff
@@ -1,112 +1,59 @@
-        help="Folder to save tokenizer to.")
+    args = sys.argv
+    config_path =args[1]
+
```
See white space comments from earlier.

JayOrten replied:
discussed

KimballNJardine commented on train_model.py:
```diff
@@ -153,20 +99,20 @@ def train_model(
-    assert dataset_dir.exists(), \
-        f"The directory with data, {dataset_dir}, doesn't exist!"
-    print(f"\nUsing dataset directory {dataset_dir}")
+    # dataset_dir = Path(datasets_dir) / dataset_name
```
Decide if we want to remove the code entirely, or leave it as comments.

KimballNJardine commented on train_model.py:
```diff
@@ -176,73 +122,73 @@ def train_model(
-    print(f"Saving tokenizer files in {tokenizers_dir}")
+    # tokenizers_dir = Path(models_path) / "tokenizers"
+    # tokenizers_dir.mkdir(parents=False, exist_ok=True)
+    # print(f"Saving tokenizer files in {tokenizers_dir}")
```
Decide if we want to remove code entirely or leave as comments.

KimballNJardine commented on train_model.py:
```diff
@@ -176,73 +122,73 @@ def train_model(
     # Save all the variables in args as JSON inside folder
     json_string = json.dump(
-        obj=arg_dict,
+        obj=arg_table,
```
Double check if arg_dict is used for anything else. If no, remove arg_dict (created at start of method.)

JayOrten replied:
It's used up above for something else.

KimballNJardine commented on train_model.py:
```diff
@@ -390,59 +336,11 @@ def train_model(
-    train_model(**vars(args))
+    args = sys.argv
+    config_path =args[1]
+
```
See earlier comment about white space.

KimballNJardine commented on train_tokenizer.py:
```diff
@@ -102,62 +95,18 @@ def train_tokenizer(
-    train_tokenizer(**vars(args))
+    args = sys.argv
+    config_path =args[1]
+
```
See earlier comments about white space.

KimballNJardine reviewed (changes requested):
This really cleans up the code. I have some initial thoughts and issues that stuck out to me, but this looks really good.

JayOrten committed:
add parantheses

JayOrten committed:
add splits comment

JayOrten committed:
update comment

JayOrten committed:
remove duplicate lines

jtappen1 committed:
Cleaned up ahead of PR

KimballNJardine committed:
Merge branch 'main' into implement_yaml

nprisbrey committed:
Add Back In New Line Characters At File End

nprisbrey committed:
Reformat template_config.yaml Example File

nprisbrey committed:
Fix model_type Format Typo In template_config.yaml

nprisbrey committed:
Fix Formatting Of download_data.py

nprisbrey committed:
Fix Formatting Of grid_search.py

nprisbrey committed:
Make Scripts One Liners

This is done because the full command is now less than 80 characters.

nprisbrey committed:
Fix Formatting Of tokenize_data.py

nprisbrey committed:
Fix Formatting Of train_tokenizer.py

nprisbrey committed:
Fix Formatting Of utils.py

nprisbrey committed:
Make Struct String Representation One Line

This is done to make the string representation more concise and match
other Python object formats. This also makes log files and print
statements take advantage of screen width as well.

nprisbrey committed:
Fix Formatting Of train_model.py

nprisbrey committed:
Fix Argument Printing And Model Args Storage Bugs

nprisbrey committed:
Improve Structure Of Struct Class

nprisbrey committed:
Simplify Redundant Model Training Scripts

Since the implementation of YAML makes the specification of which model
type to train in the config file, having two different model training
scripts that are now exactly the same is redundant.

nprisbrey reviewed (approved):
I added some tweaks in the latest commits. As always, feel free to push back against them if anything doesn't seem correct. Now waiting for @KimballNJardine's approval.

KimballNJardine commented on scripts/train_tokenizer.sh:
Note comment for scripts/download_data.sh below.

JayOrten replied:
changed

KimballNJardine commented on scripts/download_data.sh:
For all the base scripts it does the following:
```python3 ../{name_of_shell_file}.py ../configs/user_configs/YOUR_CONFIG_HERE.yaml```.

This file structure implicitly suggests that it is being called while the user's working directory is ```/scripts```.
However, user scripts should be stored in ```/scripts/user_scripts```.
Going up one level would be in ```/scripts```, not the top level directory, so neither ```../{name_of_shell_file}.py``` nor  ```../configs/user_configs/YOUR_CONFIG_HERE.yaml``` could be found.

**I recommend changing all sh files to the following:
```python3 ../../{name_of_shell_file}.py ../../configs/user_configs/YOUR_CONFIG_HERE.yaml```
(Note the extra ```../``` on each part of the sh script)**

If I have misunderstood the file directories, please let me know. I am also willing to drop the point if everyone envisions calling the script another way, like ```sh user_scripts/download_data.sh``` while cwd is ```/scripts```.

JayOrten replied:
Good point, I agree. I will make those changes

JayOrten replied:
changed

KimballNJardine commented on scripts/grid_search.sh:
Note comment for scripts/download_data.sh above.

KimballNJardine commented on scripts/tokenize_data.sh:
Note comment for scripts/download_data.sh above.

KimballNJardine commented on scripts/train_model.sh:
Note comment for scripts/download_data.sh above.

KimballNJardine reviewed (changes requested):
I have a minor request regarding the .sh files. It does not need to be changed per say, but it needs to be discussed. Otherwise, good.

JayOrten committed:
update scripts paths

KimballNJardine reviewed (approved):
Looks good to me.
